modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java

```diff
@@ -0,0 +1,531 @@
+import java.lang.reflect.InvocationTargetException;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static java.lang.invoke.MethodHandles.Lookup;
+import static org.elasticsearch.painless.Compiler.Loader;
+import static org.elasticsearch.painless.WriterConstants.CLASS_VERSION;
+import static org.elasticsearch.painless.WriterConstants.DELEGATE_BOOTSTRAP_HANDLE;
+import static org.objectweb.asm.Opcodes.ACC_FINAL;
+import static org.objectweb.asm.Opcodes.ACC_PRIVATE;
+import static org.objectweb.asm.Opcodes.ACC_PUBLIC;
+import static org.objectweb.asm.Opcodes.ACC_STATIC;
+import static org.objectweb.asm.Opcodes.ACC_SUPER;
+import static org.objectweb.asm.Opcodes.ACC_SYNTHETIC;
+import static org.objectweb.asm.Opcodes.H_INVOKEINTERFACE;
+import static org.objectweb.asm.Opcodes.H_INVOKESTATIC;
+import static org.objectweb.asm.Opcodes.H_INVOKEVIRTUAL;
+import static org.objectweb.asm.Opcodes.H_NEWINVOKESPECIAL;
+
+/**
+ * LambdaBootstrap is used to generate all the code necessary to execute
+ * lambda functions and method references within Painless.  The code generation
+ * used here is based upon the following article:
+ * http://cr.openjdk.java.net/~briangoetz/lambda/lambda-translation.html
+ * However, it is a simplified version as Painless has no concept of generics.
+ * LambdaBootstrap is being used as a replacement for
+ * {@link java.lang.invoke.LambdaMetafactory} since the Painless casting model
+ * cannot be fully supported through this class.
+ *
+ * For each lambda function/method reference used within a Painless script
+ * a class will be generated at link-time using the
+ * {@link LambdaBootstrap#lambdaBootstrap} method that contains the following:
+ * 1. member fields for any captured variables
+ * 2. a constructor that will take in captured variables and assign them to
+ * their respective member fields
+ * 3. if there are captures, a factory method that will take in captured
+ * variables and delegate them to the constructor
+ * 4. a method that will load the member fields representing captured variables
+ * and take in any other necessary values based on the arguments passed into the
+ * lambda function/reference method; it will then make a delegated call to the
+ * actual lambda function/reference method
+ *
+ * Take for example the following Painless script:
+ *
+ * {@code
+ * List list1 = new ArrayList(); "
+ * list1.add(2); "
+ * List1 list2 = new ArrayList(); "
+ * list1.forEach(x -> list2.add(x));"
+ * return list[0]"
+ * }
+ *
+ * The script contains a lambda function with a captured variable.
+ * The following Lambda class would be generated:
+ *
+ * {@code
+ *     public static final class $$Lambda0 implements Consumer {
+ *         private List arg$0;
+ *
+ *         public $$Lambda0(List arg$0) {
+ *             this.arg$0 = arg$0;
+ *         }
+ *
+ *         public static $$Lambda0 get$Lambda(List arg$0) {
```

Comment:
To invoke a constructor the class would have to be newed and duped onto the stack.  The caller doesn't know what the generated class is, so it calls a factorymethod that can create the class instead.